File: packages/ts-sdk/src/treedx/client/pagination.ts
import type { TreeDxPage } from '../types/index.js';

export function createPage<T>(input: TreeDxPage<T>): TreeDxPage<T> {
  return {
    items: input.items,
    nextCursor: input.nextCursor,
    hasMore: input.hasMore,
    cursor: input.cursor,
    limit: input.limit
  };
}

export function isTreeDxPage(value: unknown): value is TreeDxPage<unknown> {
  return typeof value === 'object' && value !== null && Array.isArray((value as TreeDxPage<unknown>).items);
}

export function getNextCursor<T>(page: TreeDxPage<T>): string | undefined {
  return page.nextCursor;
}

File: packages/python-sdk/src/treedx_sdk/pagination.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Generic, TypeAlias, TypeVar


T = TypeVar("T")
TreeDxCursor: TypeAlias = str


@dataclass(frozen=True)
class TreeDxPage(Generic[T]):
    items: list[T]
    next_cursor: str | None = None
    has_more: bool | None = None
    cursor: str | None = None
    limit: int | None = None


def create_page(
    items: list[T],
    *,
    next_cursor: str | None = None,
    has_more: bool | None = None,
    cursor: str | None = None,
    limit: int | None = None,
) -> TreeDxPage[T]:
    return TreeDxPage(items=items, next_cursor=next_cursor, has_more=has_more, cursor=cursor, limit=limit)


def is_treedx_page(value: Any) -> bool:
    return isinstance(value, TreeDxPage) or (isinstance(value, dict) and isinstance(value.get("items"), list))


def get_next_cursor(page: TreeDxPage[Any]) -> str | None:
    return page.next_cursor
